system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,407 @@
+                        if let Some(domain_name) = domain_name.downcast::<CFString>() {
+                            _domain_name = Some(domain_name.to_string());
+                        }
+                    }
+
+                    if let Some(addrs) =
+                        dict.find2(&CFString::from_static_string("ServerAddresses"))
+                    {
+                        let addrs = unsafe { CFType::wrap_under_get_rule(addrs) };
+                        if let Some(addrs) = addrs.downcast::<CFArray<CFString>>() {
+                            let mut temp = Vec::new();
+                            for addr in addrs.iter() {
+                                if let Ok(ip_addr) = addr.to_string().parse::<IpAddr>() {
+                                    temp.push(ip_addr);
+                                }
+                            }
+
+                            if temp.len() > 0 {
+                                _server_addresses = Some(temp);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return (_domain_name, _server_addresses);
+        };
+
+        let (state_domain_name, state_server_addresses) =
+            query(format!("State:/Network/Service/{}/DNS", self.id()));
+        let (setup_domain_name, setup_server_addresses) =
+            query(format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+        SCNetworkServiceDNS {
+            state_domain_name: state_domain_name,
+            state_server_addresses: state_server_addresses,
+            setup_domain_name: setup_domain_name,
+            setup_server_addresses: setup_server_addresses,
+        }
+    }
+
+    pub fn set_dns(&self, dns: SCNetworkServiceDNS) -> bool {
+        let store = SCDynamicStoreBuilder::new("ns_dns_set").build();
+
+        if dns.setup_server_addresses.is_some() {
+            let key = CFString::from_static_string("ServerAddresses");
+            let addrs: Vec<CFString> = dns.setup_server_addresses
+                .unwrap()
+                .iter()
+                .map(|s| CFString::new(&format!("{}", s)))
+                .collect();
+            let value = CFArray::from_CFTypes(&addrs);
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                return false;
+            }
+        }
+
+        if dns.setup_domain_name.is_some() {
+            let key = CFString::from_static_string("DomainName");
+            let value = CFString::new(dns.setup_domain_name.unwrap().as_str());
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                // FIXME: should rollback ?
```

Comment:
It seems that `SCDynamicStoreSetMultiple` would be more appropriate for setting multiple values.

@faern is it implemented in the store?

Reply:
Nope, we have not implemented a safe abstraction for `SetMultiple` yet, no. But this method only sets one value, doesn't it?

Reply:
set `ServerAddresses` and `DomainName` .